### DISCUSSION TITLE
Other backends for debug_printf()

### DISCUSSION THREAD

**Discussion by dresco** [grblHAL/core#545] [General]
Thanks for adding the additional serial stream debug output options.

It would be useful to add support for debugger interfaces like SWO or Segger RTT. These would normally be written with printf(), with driver code to override the output - like [this](https://github.com/dresco/STM32H7xx/blob/11744290e1a8cefd0962ec4e8d554c83257aeb2f/Src/swo.c#L27-L38) for SWO.

Would it be possible to add a similar debug_printf() implementation that targets printf() when enabled, and a NOP if DEBUG is not defined?

**Comment by terjeio** [comment]
Thinking aloud...

One way could be to use the `DEBUGOUT` symbol to direct the output? E.g. add something like this [here](https://github.com/grblHAL/core/blob/5cff12e6608b1f53e4becff4e853e460fc1b7cbf/stream.c#L652):

```
#ifdef DEBUGOUT

#if DEBUGOUT == -1

#include "core_cm4.h"

void debug_write (const char *s)
{
    char c;

    while((c = *s++))
        ITM_SendChar(c);
}

void debug_writeln (const char *s)
{
    debug_write(s);
    debug_write(ASCII_EOL);
}

void debug_printf (const char *fmt, ...)
{
    char debug_out[DEBUG_BUFFER];

    va_list args;
    va_start(args, fmt);
    vsnprintf(debug_out, sizeof(debug_out) - 1, fmt, args);
    va_end(args);

    debug_writeln(debug_out);
}

bool debug_stream_init (void)
{
    return true;
}

#else

static stream_write_ptr dbg_write = NULL;
...
```
Perhaps make `debug_write` a weak function that could be overridden in a more appropriate location (out of the core since availability will be board specific)?
Another option could be adding the output as a grblHAL stream and not modify the core?

**Reply by dresco** [reply]
I quite like the idea of making `debug_write` (or even `debug_printf`) a weak function that could be overriden locally in the driver.

Both the SWO and Segger RTT implementations already override the low level GCC `_write` function, so `debug_printf` could just call `printf` directly for either of these? Don't know whether there is actually any benefit to bypassing the debug_out buffer etc, just a thought...

**Reply by terjeio** [reply]
I have now committed the code above with debug_write() as a weak function.

> Both the SWO and Segger RTT implementations already override the low level GCC _write function, so debug_printf could just call printf directly for either of these?

Will this be true for all supported platforms/processors? When this is confirmed then such a change may make sense.

> Don't know whether there is actually any benefit to bypassing the debug_out buffer etc, just a thought...

Having the buffer allows for setting a breakpoint in debug_printf () and inspect the buffer if no output is available?

**Reply by dresco** [reply]
> I have now committed the code above with debug_write() as a weak function.

Thanks, will migrate my scattered `printf()` calls across..

> Will this be true for all supported platforms/processors? 

Good question. Believe SWO is a standard feature on Cortex M3, M4, & M7, although I have only used on ST parts.

Is not available on M0 though, so for those (or for other boards where the SWO pin is already used), Segger RTT can be used - does require a JLink adapter though. I have one here, so will give both options a sanity check..

**Comment by dresco** [comment]
Hi @terjeio 

Have tested locally on the H7 driver.  Is working well over SWO/RTT, and when DEBUGOUT is defined to a spare serial port.

However, I get a hardfault when console is on USB connection and DEBUG is defined without DEBUGOUT. It seems get_tx_buffer_count() is null for the USB stream?

https://github.com/grblHAL/core/blob/8b9b7859ee6b370442d380d3d754fa5392b064b2/stream.c#L764

**Reply by terjeio** [reply]
Oops, I tested with my Nucleo-64 board and that does not have USB...
I'll check for `hal.stream.get_tx_buffer_count != NULL` in the next build.

**Comment by dresco** [comment]
> I'll check for hal.stream.get_tx_buffer_count != NULL in the next build.

Thanks, I've pushed an [update](https://github.com/dresco/STM32H7xx/commit/fe859dce9cb3c9a713c1eb5e0a48e1e798b5a4ab#diff-1a978c2ebcda6fd189cefce17f4f040df33621e8cca4d9c334e186cb83ab7250) to the H7 driver. Should also work as-is for F3, F4, F7..

Defaults to SWO, but can be overriden to use Segger RTT with a `DEBUGOUT_BACKEND` definition. Also overrides the printf() backend in both cases, as noticed a few libraries where that may be used for log output (LwIP, littlefs, USB etc).
